ttnn/cpp/ttnn/tensor/vector_base.cpp

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "vector_base.hpp"
+#include <stdexcept>
+#include "fmt/color.h"
+#include "tt_metal/common/assert.hpp"
+
+namespace tt::tt_metal {
+
+namespace {
+
+int32_t normalized_index(int32_t index, size_t original_size, size_t container_size) {
+    int32_t orig_size = static_cast<int32_t>(original_size);
+    int32_t full_size = static_cast<int32_t>(container_size);
+
+    int fixed_index = index;
+    if (fixed_index < 0) {
+        fixed_index += full_size;
+    } else {
+        fixed_index += full_size - orig_size;
+    }
+
+    if (fixed_index < 0 || fixed_index >= full_size) {
+        TT_THROW("VectorBase[] index out of range. {} not in [{}, {})", index, -full_size, full_size);
+    }
+
+    return fixed_index;
+}
+}
+
+void VectorBase::init() {
+    m_original_size = m_value.size();
+    const size_t min_internal_size = 4;
+
+    if(m_original_size < min_internal_size) {
+        std::vector<uint32_t> ones(min_internal_size - m_original_size, 1);
```

Comment:
This would cause an extra allocation, you can do it like this for example:
```
if(m_original_size < MIN_INTERNAL_SIZE) {
        m_value.resize(MIN_INTERNAL_SIZE);
        size_t shift = MIN_INTERNAL_SIZE - m_original_size;
        for (size_t idx = MIN_INTERNAL_SIZE - 1; idx >= shift; idx--) {
            m_value[idx] = m_value[idx - shift];
        }
        for(size_t idx = 0; idx < shift; idx++) {
            m_value[idx] = 1;
        }
    }
```